src/shared_data.rs

```diff
@@ -0,0 +1,237 @@
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::{utils, PublicKey, Result, XorName};
+use multibase::Decodable;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use std::ops::Range;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct Key(Vec<u8>);
+
+impl Key {
+    pub fn get(&self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+}
+
+impl Deref for Key {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Vec<u8>> for Key {
+    fn from(vec: Vec<u8>) -> Self {
+        Key(vec)
+    }
+}
+
+// pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+pub type KvPair = (Key, Value);
+pub type Values = Vec<Value>;
+pub type Keys = Vec<Key>;
+
+/// Marker for sentried data.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct Sentried;
+
+/// Marker for non-sentried data.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct NonSentried;
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub enum User {
+    Anyone,
+    Specific(PublicKey),
+}
+
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum Version {
+    FromStart(u64), // Absolute index
+    FromEnd(u64),   // Relative Version - start counting from the end
+}
+
+impl From<u64> for Version {
+    fn from(version: u64) -> Self {
+        Version::FromStart(version)
+    }
+}
+
+// Set of data, owners, permissions versions.
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct ExpectedVersions {
+    expected_data_version: u64,
+    expected_owners_version: u64,
+    expected_access_list_version: u64,
+}
+
+impl ExpectedVersions {
+    pub fn new(
+        expected_data_version: u64,
+        expected_owners_version: u64,
+        expected_access_list_version: u64,
+    ) -> Self {
+        ExpectedVersions {
+            expected_data_version,
+            expected_owners_version,
+            expected_access_list_version,
+        }
+    }
+
+    pub fn expected_data_version(&self) -> u64 {
+        self.expected_data_version
+    }
+
+    pub fn expected_owners_version(&self) -> u64 {
+        self.expected_owners_version
+    }
+
+    pub fn expected_access_list_version(&self) -> u64 {
+        self.expected_access_list_version
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct Owner {
+    pub public_key: PublicKey,
+    /// The expected Version of the data at the time this ownership change is to become valid.
+    pub expected_data_version: u64,
```

Comment:
If I understand your question correctly: no.

Except for naming, the implementation is the same now as before. I have also thought slightly about if something should change here, but it went up fairly low on my priority list.

The Owner struct is used both as a DTO and as a domain model, which is a bit iffy IMO, because it is used for slightly different reasons, which are not apparent (as evidenced by your question as well).

As a DTO in a request to set the owner, the `expected_data_version` and `expected_access_list_version` acts as optimistic concurrency control, i.e. if anything changed in the data or permissions from the point that the requester sees it, to when the request is handled, it is rejected.
As a domain model, it tells the history of who the owner was at a particular data version. This would presumably let us answer questions like: `when current data version was x, who was the owner? when this set of users had this set of permissions, who was the owner?`